Format called and returned with args

qha opened the pull request:
This pr closes #155 and Olivine-Labs/busted#578 and replaces #164.

qha commented:
Obviously there is another outstanding issue too. It may be a week or two until i can resolve it.

qha commented:
Nope, found it.

qha commented:
I figured out how to check messages for failed assertions so the formatting questions are the only ones that remain, see above.

dav-is commented:
Is there an update on this PR?

qha commented:
> Is there an update on this PR?
I presume this is directed at someone else but just in case i'm mistaken: I'm waiting for feedback on the outstanding issues still listed in the top comment or any other feedback on changes needed to get this merged.

Tieske commented on spec/formatters_spec.lua:
```diff
@@ -170,5 +189,8 @@ describe("Test Formatters", function()
     assert:remove_formatter(s)
   end)
-
 end)
```
can you please remove unrelated changes like this? It clutters the diff and makes it hard(er) to review

qha replied:
Gone.

Tieske commented on spec/formatters_spec.lua:
```diff
@@ -170,5 +189,8 @@ describe("Test Formatters", function()
+
+-- Local variables:
+-- lua-indent-level: 2;
+-- end:
```
seems like some tooling related thing. Can you please remove this?

qha replied:
Gone.

Tieske commented on spec/matchers_spec.lua:
```diff
@@ -288,12 +288,12 @@ describe("Test Matchers", function()
+    assert.spy(s).returned_with(match.is_not_number())
+    assert.spy(s).returned_with(match.is_not_table())
+    assert.spy(s).has_not.returned_with(match.is_number())
+    assert.spy(s).has_not.returned_with(match.is_table())
```
there are many of these types changes. Are they related? seems not. Can you please remove those?

qha replied:
Gone.

Tieske commented on src/util.lua:
```diff
@@ -100,42 +104,89 @@ function util.copyargs(args)
+-- @param arglist the table to clear of arguments or return values and their count
+-- @return No return values
+function util.cleararglist(arglist)
+  for idx = #arglist, 1, -1 do
```
`#arglist` will fail if there are `nil`s in the table. Should be a numeric loop from `arglist.n`

qha replied:
Updated.

Tieske commented on src/util.lua:
```diff
@@ -100,42 +104,89 @@ function util.copyargs(args)
+  -- your assertions between calls to check this gives you the best
+  -- information if no calls match. In the cases where you can't do
+  -- that there is no good way to predict what would work best.
+  for ii = #argslist, 1, -1 do
```
fails on `nil`s

qha replied:
Updated.

Tieske replied:
how about `invocations`

Tieske commented on src/util.lua:
```diff
@@ -283,4 +334,19 @@ function util.extract_keys(nspace, tokens)
+end
+
+function util.is_arglist(object)
+  return type(object) == "table" and getmetatable(object) == arglist_mt
```
the first check (for being a table) is not necessary, every type has a metatable

qha replied:
I probably copied this from a similar function in spy.lua or match.lua but will remove it in an updated commit.

qha commented:
Thank you for your feedback! I will have a look at addressing your comments. I don't expect to find the time for that this week though, hopefully next week or the week after that.

qha commented:
I found some time today for working on this pr.

Please review the updated commits and my comments in the conversations above!

jeduden commented:
Looks like a fantastic feature to have. Is there any chance that this PR will be merged ?

Tieske commented:
merged manually
